substrate/codec/src/slicable.rs

```diff
@@ -87,25 +87,75 @@ impl<T: Slicable, E: Slicable> Slicable for Result<T, E> {
 	}
 }
 
-impl Slicable for Option<bool> {
+/// Shim type because we can't do a specialised implementation for `Option<bool>` directly.
+pub struct OptionBool (Option<bool>);
+
+impl Slicable for OptionBool {
 	fn decode<I: Input>(input: &mut I) -> Option<Self> {
 		match input.read_byte()? {
-			0 => Some(Some(false)),
-			1 => Some(Some(true)),
-			2 => Some(None),
+			0 => Some(OptionBool(None)),
+			1 => Some(OptionBool(Some(true))),
+			2 => Some(OptionBool(Some(false))),
 			_ => None,
 		}
 	}
 
 	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
 		f(&[match *self {
-			Some(false) => 0u8,
-			Some(true) => 1u8,
-			None => 2u8,
+			OptionBool(None) => 0u8,
+			OptionBool(Some(true)) => 1u8,
+			OptionBool(Some(false)) => 2u8,
 		}])
 	}
 }
 
+impl<T: Slicable> Slicable for Option<T> {
+	fn decode<I: Input>(input: &mut I) -> Option<Self> {
+		match input.read_byte()? {
+			0 => Some(None),
+			1 => Some(Some(T::decode(input)?)),
+			_ => None,
+		}
+	}
+
+	fn encode(&self) -> Vec<u8> {
+		let mut v = Vec::new();
+		match *self {
+			Some(ref t) => {
+				v.push(1);
+				t.using_encoded(|s| v.extend(s));
+			}
+			None => v.push(0),
+		}
+		v
+	}
+}
+
+macro_rules! impl_array {
+	( $( $n:expr )* ) => { $(
+		impl<T: Slicable> Slicable for [T; $n] {
+			fn decode<I: Input>(input: &mut I) -> Option<Self> {
+				let mut r: Self = unsafe { mem::uninitialized() };
+				for i in 0..$n {
+					r[i] = T::decode(input)?;
```

Comment:
A bunch of serious issues with this code. If `T` has a destructor:
  - writing each element without using ptr::write will attempt to drop uninitialized memory (this is UB)
  - if the decode implementation panics before all array members are filled, it will attempt to drop the uninitialized members as well. this is UB.

We should just bound by `Default`. I really strongly recommend against using unsafe code until it's proven to be necessary and we always seem to get it subtly wrong.

Reply:
Default is just horrible.